src/Ormolu/Utils.hs

```diff
@@ -150,3 +157,17 @@ matchAddEpAnn :: AnnKeywordId -> AddEpAnn -> Maybe EpaLocation
 matchAddEpAnn annId (AddEpAnn annId' loc)
   | annId == annId' || unicodeAnn annId == annId' = Just loc
   | otherwise = Nothing
+
+-- | Convert 'Text' to a 'StringBuffer' by making a copy.
+textToStringBuffer :: Text -> StringBuffer
+textToStringBuffer txt = unsafePerformIO $ do
+  buf <- mallocPlainForeignPtrBytes (len + 3)
+  withForeignPtr buf $ \ptr -> do
+    TFFI.unsafeCopyToPtr txt ptr
+    -- last three bytes have to be zero for easier decoding
+    pokeElemOff ptr len 0
+    pokeElemOff ptr (len + 1) 0
+    pokeElemOff ptr (len + 2) 0
+  pure StringBuffer {buf, len, cur = 0}
+  where
+    len = TFFI.lengthWord8 txt
```

Comment:
If desired, this could also be written by using `ByteString` as an intermediary type (which is then more obviously correct), but it would incur two copies instead of just one here.

Reply:
I am okay with this implementation, but perhaps `unsafePerformIO` is not necessary? Let it live in `IO` and we can plug in it at some point in `ormolu`, since it already has the `MonadIO m` constraint. Functions called from `ormolu` and its variants could take `StringBuffer` directly instead of accepting `Text` and re-converting it.

Reply:
Hmm, that sounds good, but we use `textToStringBuffer` e.g. in the pure `tokenize` function `Ormolu.Parse.Pragma`, and also manipulate the input via `Data.Text`, so don't see an easy way to both keep the functions there pure and avoid converting back from `StringBuffer` to `Text`.

Reply:
Yeah, I don't have a good idea how to avoid `IO` in that context.

Reply:
[`stringBufferFromByteString`](https://hackage.haskell.org/package/ghc-lib-parser-9.4.4.20221225/docs/GHC-Data-StringBuffer.html#v:stringBufferFromByteString) is defined next to `textToStringBuffer` and doing the same under the hood, but doesn't expose IO.

I have an idea to avoid copying at all.
StringBuffer is wrapper around ForeignPtr:

```
data StringBuffer
 = StringBuffer {
     buf :: {-# UNPACK #-} !(ForeignPtr Word8),
     len :: {-# UNPACK #-} !Int,        -- length
     cur :: {-# UNPACK #-} !Int         -- current pos
  }
```
there is a [mmap](https://hackage.haskell.org/package/mmap-0.5.9/docs/System-IO-MMap.html) library which can mmap file and produce ForeignPtr:
```
System.IO.MMap.mmapFileForeignPtr :: FilePath                     -- ^ name of file to map
                   -> Mode                         -- ^ access mode
                   -> Maybe (Int64,Int)            -- ^ range to map, maps whole file if Nothing
                   -> IO (ForeignPtr a,Int,Int)
```

Let's mmap file in read/write mode with size + 3 bytes and zero last 3 bytes.

In case of shared memory causes exception then fallback to current approach with copying.

Reply:
Writing zero must not be required because OS should take care of zeroing new blocks because of security.

Reply:
Thanks for the suggestion, but I don't think it is possible to make this work, as we need those three extra zero bytes at the end of the `ForeignPtr`, which does not seem to be possible via `mmapFileForeignPtr` (we don't want to actually modify the file).